File: src/Core/Table/CompositeCell.fs
namespace ARCtrl

open Fable.Core

[<AttachMembers>]
[<RequireQualifiedAccess>]
type CompositeCell = 
    /// ISA-TAB term columns as ontology annotation.
    ///
    /// https://isa-specs.readthedocs.io/en/latest/isatab.html#ontology-annotations
    | Term of OntologyAnnotation
    /// Single columns like Input, Output, ProtocolREF, .. .
    | FreeText of string
    /// ISA-TAB unit columns, consisting of value and unit as ontology annotation.
    ///
    /// https://isa-specs.readthedocs.io/en/latest/isatab.html#unit
    | Unitized of string*OntologyAnnotation
    | Data of Data

    member this.isUnitized = match this with | Unitized _ -> true | _ -> false
    member this.isTerm = match this with | Term _ -> true | _ -> false
    member this.isFreeText = match this with | FreeText _ -> true | _ -> false
    member this.isData = match this with | Data _ -> true | _ -> false

    /// <summary>
    /// This returns the default empty cell from an existing CompositeCell.
    /// </summary>
    member this.GetEmptyCell() =
        match this with
        | CompositeCell.Term _ -> CompositeCell.emptyTerm
        | CompositeCell.Unitized _ -> CompositeCell.emptyUnitized
        | CompositeCell.FreeText _ -> CompositeCell.emptyFreeText
        | CompositeCell.Data _ -> CompositeCell.emptyData

    /// <summary>
    /// This function returns an array of all values as string
    ///
    /// ```fsharp
    /// match this with
    /// | FreeText s -> [|s|]
    /// | Term oa -> [| oa.NameText; defaultArg oa.TermSourceREF ""; defaultArg oa.TermAccessionNumber ""|]
    /// | Unitized (v,oa) -> [| v; oa.NameText; defaultArg oa.TermSourceREF ""; defaultArg oa.TermAccessionNumber ""|]
    /// ```
    /// </summary>
    member this.GetContent() = 
        match this with
        | FreeText s -> [|s|]
        | Term oa -> [| oa.NameText; defaultArg oa.TermSourceREF ""; defaultArg oa.TermAccessionNumber ""|]
        | Unitized (v,oa) -> [| v; oa.NameText; defaultArg oa.TermSourceREF ""; defaultArg oa.TermAccessionNumber ""|]
        | Data d -> [| defaultArg d.Name ""; defaultArg d.Format ""; defaultArg d.SelectorFormat ""|]

    /// FreeText string will be converted to unit term name.
    ///
    /// Term will be converted to unit term.
    member this.ToUnitizedCell() =
        match this with
        | Unitized _ -> this
        | FreeText text -> CompositeCell.Unitized ("", OntologyAnnotation.create(text))
        | Term term -> CompositeCell.Unitized ("", term)
        | Data d -> CompositeCell.Unitized ("", OntologyAnnotation.create(d.NameText))

    /// FreeText string will be converted to term name.
    ///
    /// Unit term will be converted to term and unit value is dropped.
    member this.ToTermCell() =
        match this with
        | Term _ -> this
        | Unitized (_,unit) -> CompositeCell.Term unit
        | FreeText text -> CompositeCell.Term(OntologyAnnotation.create(text))
        | Data d -> CompositeCell.Term(OntologyAnnotation(d.NameText))

    /// Will always keep `OntologyAnnotation.NameText` from Term or Unit.
    member this.ToFreeTextCell() =
        match this with
        | FreeText _ -> this
        | Term term -> FreeText(term.NameText)
        | Unitized (_,unit) -> FreeText(unit.NameText)
        | Data d -> FreeText (d.NameText)

    member this.ToDataCell() =
        match this with
        | Unitized (_, unit) -> CompositeCell.createDataFromString unit.NameText
        | FreeText txt -> CompositeCell.createDataFromString txt
        | Term term -> CompositeCell.createDataFromString term.NameText
        | Data _ -> this

    // Suggest this syntax for easy "of-something" access
    member this.AsUnitized  =
        match this with
        | Unitized (v,u) -> v,u
        | _ -> failwith "Not a Unitized cell."

    member this.AsTerm =
        match this with
        | Term c -> c
        | _ -> failwith "Not a Term Cell."

    member this.AsFreeText =
        match this with
        | FreeText c -> c
        | _ -> failwith "Not a FreeText Cell."

    member this.AsData =
        match this with
        | Data d -> d
        | _ -> failwith "Not a Data Cell."

    // TODO: i would really love to have an overload here accepting string input
    static member createTerm (oa:OntologyAnnotation) = Term oa

    static member createTermFromString (?name: string, ?tsr: string, ?tan: string) =
        Term <| OntologyAnnotation.create(?name = name, ?tsr = tsr, ?tan = tan)

    static member createUnitized (value: string, ?oa:OntologyAnnotation) = Unitized (value, Option.defaultValue (OntologyAnnotation()) oa)

    static member createUnitizedFromString (value: string, ?name: string, ?tsr: string, ?tan: string) = 
        Unitized <| (value, OntologyAnnotation.create(?name = name, ?tsr = tsr, ?tan = tan))

    static member createFreeText (value: string) = FreeText value
    
    static member createData (d:Data) = Data d

    static member createDataFromString (value : string, ?format : string, ?selectorFormat : string) =
        Data(Data.create(Name = value, ?Format = format, ?SelectorFormat = selectorFormat))

    static member emptyTerm = Term (OntologyAnnotation())
    static member emptyFreeText = FreeText ""
    static member emptyUnitized = Unitized ("", OntologyAnnotation())
    static member emptyData = Data(Data.create())

    /// <summary>
    /// Updates current CompositeCell with information from OntologyAnnotation.
    ///
    /// For `Term`, OntologyAnnotation (oa) is fully set. For `Unitized`, oa is set as unit while value is untouched.
    /// For `FreeText` oa.NameText is set.
    /// </summary>
    /// <param name="oa"></param>
    member this.UpdateWithOA (oa:OntologyAnnotation) =
        match this with
        | CompositeCell.Term _ -> CompositeCell.createTerm oa
        | CompositeCell.Unitized (v,_) -> CompositeCell.createUnitized (v,oa)
        | CompositeCell.FreeText _ -> CompositeCell.createFreeText oa.NameText
        | CompositeCell.Data d ->
            d.Name <- Some oa.NameText
            CompositeCell.Data d

    /// <summary>
    /// Updates current CompositeCell with information from OntologyAnnotation.
    ///
    /// For `Term`, OntologyAnnotation (oa) is fully set. For `Unitized`, oa is set as unit while value is untouched.
    /// For `FreeText` oa.NameText is set.
    /// </summary>
    /// <param name="oa"></param>
    /// <param name="cell"></param>
    static member updateWithOA (oa:OntologyAnnotation) (cell: CompositeCell) =
        cell.UpdateWithOA oa

    override this.ToString() = 
        match this with
        | Term oa -> $"{oa.NameText}"
        | FreeText s -> s
        | Unitized (v,oa) -> $"{v} {oa.NameText}"
        | Data d -> $"{d.NameText}"

    member this.Copy() =
        match this with
        | Term oa -> Term (oa.Copy())
        | FreeText s -> FreeText s
        | Unitized (v,oa) -> Unitized (v, oa.Copy())
        | Data d -> Data (d.Copy())

#if FABLE_COMPILER
    //[<CompiledName("Term")>]
    static member term (oa:OntologyAnnotation) = CompositeCell.Term(oa)

    //[<CompiledName("FreeText")>]
    static member freeText (s:string) = CompositeCell.FreeText(s)

    //[<CompiledName("Unitized")>]
    static member unitized (v:string, oa:OntologyAnnotation) = CompositeCell.Unitized(v, oa)

    //[<CompiledName("Data")>]
    static member data (d:Data) = CompositeCell.Data(d)
#else
#endif

File: js/Core/Table/CompositeCell.js
import { unwrap, defaultArg } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { OntologyAnnotation_$reflection, OntologyAnnotation } from "../OntologyAnnotation.js";
import { Data_$reflection, Data } from "../Data.js";
import { Union } from "../../fable_modules/fable-library-js.4.22.0/Types.js";
import { union_type, string_type } from "../../fable_modules/fable-library-js.4.22.0/Reflection.js";

export class CompositeCell extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Term", "FreeText", "Unitized", "Data"];
    }
    get isUnitized() {
        const this$ = this;
        return this$.tag === 2;
    }
    get isTerm() {
        const this$ = this;
        return this$.tag === 0;
    }
    get isFreeText() {
        const this$ = this;
        return this$.tag === 1;
    }
    get isData() {
        const this$ = this;
        return this$.tag === 3;
    }
    GetEmptyCell() {
        const this$ = this;
        return (this$.tag === 2) ? CompositeCell.emptyUnitized : ((this$.tag === 1) ? CompositeCell.emptyFreeText : ((this$.tag === 3) ? CompositeCell.emptyData : CompositeCell.emptyTerm));
    }
    GetContent() {
        const this$ = this;
        switch (this$.tag) {
            case 0: {
                const oa = this$.fields[0];
                return [oa.NameText, defaultArg(oa.TermSourceREF, ""), defaultArg(oa.TermAccessionNumber, "")];
            }
            case 2: {
                const oa_1 = this$.fields[1];
                return [this$.fields[0], oa_1.NameText, defaultArg(oa_1.TermSourceREF, ""), defaultArg(oa_1.TermAccessionNumber, "")];
            }
            case 3: {
                const d = this$.fields[0];
                return [defaultArg(d.Name, ""), defaultArg(d.Format, ""), defaultArg(d.SelectorFormat, "")];
            }
            default:
                return [this$.fields[0]];
        }
    }
    ToUnitizedCell() {
        const this$ = this;
        return (this$.tag === 1) ? (new CompositeCell(2, ["", OntologyAnnotation.create(this$.fields[0])])) : ((this$.tag === 0) ? (new CompositeCell(2, ["", this$.fields[0]])) : ((this$.tag === 3) ? (new CompositeCell(2, ["", OntologyAnnotation.create(this$.fields[0].NameText)])) : this$));
    }
    ToTermCell() {
        const this$ = this;
        return (this$.tag === 2) ? (new CompositeCell(0, [this$.fields[1]])) : ((this$.tag === 1) ? (new CompositeCell(0, [OntologyAnnotation.create(this$.fields[0])])) : ((this$.tag === 3) ? (new CompositeCell(0, [new OntologyAnnotation(this$.fields[0].NameText)])) : this$));
    }
    ToFreeTextCell() {
        const this$ = this;
        return (this$.tag === 0) ? (new CompositeCell(1, [this$.fields[0].NameText])) : ((this$.tag === 2) ? (new CompositeCell(1, [this$.fields[1].NameText])) : ((this$.tag === 3) ? (new CompositeCell(1, [this$.fields[0].NameText])) : this$));
    }
    ToDataCell() {
        const this$ = this;
        return (this$.tag === 1) ? CompositeCell.createDataFromString(this$.fields[0]) : ((this$.tag === 0) ? CompositeCell.createDataFromString(this$.fields[0].NameText) : ((this$.tag === 3) ? this$ : CompositeCell.createDataFromString(this$.fields[1].NameText)));
    }
    get AsUnitized() {
        const this$ = this;
        if (this$.tag === 2) {
            return [this$.fields[0], this$.fields[1]];
        }
        else {
            throw new Error("Not a Unitized cell.");
        }
    }
    get AsTerm() {
        const this$ = this;
        if (this$.tag === 0) {
            return this$.fields[0];
        }
        else {
            throw new Error("Not a Term Cell.");
        }
    }
    get AsFreeText() {
        const this$ = this;
        if (this$.tag === 1) {
            return this$.fields[0];
        }
        else {
            throw new Error("Not a FreeText Cell.");
        }
    }
    get AsData() {
        const this$ = this;
        if (this$.tag === 3) {
            return this$.fields[0];
        }
        else {
            throw new Error("Not a Data Cell.");
        }
    }
    static createTerm(oa) {
        return new CompositeCell(0, [oa]);
    }
    static createTermFromString(name, tsr, tan) {
        return new CompositeCell(0, [OntologyAnnotation.create(unwrap(name), unwrap(tsr), unwrap(tan))]);
    }
    static createUnitized(value, oa) {
        return new CompositeCell(2, [value, defaultArg(oa, new OntologyAnnotation())]);
    }
    static createUnitizedFromString(value, name, tsr, tan) {
        const tupledArg = [value, OntologyAnnotation.create(unwrap(name), unwrap(tsr), unwrap(tan))];
        return new CompositeCell(2, [tupledArg[0], tupledArg[1]]);
    }
    static createFreeText(value) {
        return new CompositeCell(1, [value]);
    }
    static createData(d) {
        return new CompositeCell(3, [d]);
    }
    static createDataFromString(value, format, selectorFormat) {
        return new CompositeCell(3, [Data.create(undefined, value, undefined, unwrap(format), unwrap(selectorFormat))]);
    }
    static get emptyTerm() {
        return new CompositeCell(0, [new OntologyAnnotation()]);
    }
    static get emptyFreeText() {
        return new CompositeCell(1, [""]);
    }
    static get emptyUnitized() {
        return new CompositeCell(2, ["", new OntologyAnnotation()]);
    }
    static get emptyData() {
        return new CompositeCell(3, [Data.create()]);
    }
    UpdateWithOA(oa) {
        const this$ = this;
        switch (this$.tag) {
            case 2:
                return CompositeCell.createUnitized(this$.fields[0], oa);
            case 1:
                return CompositeCell.createFreeText(oa.NameText);
            case 3: {
                const d = this$.fields[0];
                d.Name = oa.NameText;
                return new CompositeCell(3, [d]);
            }
            default:
                return CompositeCell.createTerm(oa);
        }
    }
    static updateWithOA(oa, cell) {
        return cell.UpdateWithOA(oa);
    }
    toString() {
        const this$ = this;
        return (this$.tag === 1) ? this$.fields[0] : ((this$.tag === 2) ? (`${this$.fields[0]} ${this$.fields[1].NameText}`) : ((this$.tag === 3) ? (`${this$.fields[0].NameText}`) : (`${this$.fields[0].NameText}`)));
    }
    Copy() {
        const this$ = this;
        return (this$.tag === 1) ? (new CompositeCell(1, [this$.fields[0]])) : ((this$.tag === 2) ? (new CompositeCell(2, [this$.fields[0], this$.fields[1].Copy()])) : ((this$.tag === 3) ? (new CompositeCell(3, [this$.fields[0].Copy()])) : (new CompositeCell(0, [this$.fields[0].Copy()]))));
    }
    ValidateAgainstHeader(header, raiseException) {
        const this$ = this;
        const raiseExeption = defaultArg(raiseException, false);
        const cell = this$;
        if (header.IsDataColumn && (cell.isData ? true : cell.isFreeText)) {
            return true;
        }
        else if (header.IsDataColumn) {
            if (raiseExeption) {
                throw new Error(`Invalid combination of header \`${header}\` and cell \`${cell}\`, Data header should have either Data or Freetext cells`);
            }
            return false;
        }
        else if (header.IsTermColumn && (cell.isTerm ? true : cell.isUnitized)) {
            return true;
        }
        else if (!header.IsTermColumn && cell.isFreeText) {
            return true;
        }
        else {
            if (raiseExeption) {
                throw new Error(`Invalid combination of header \`${header}\` and cell \`${cell}\``);
            }
            return false;
        }
    }
    static term(oa) {
        return new CompositeCell(0, [oa]);
    }
    static freeText(s) {
        return new CompositeCell(1, [s]);
    }
    static unitized(v, oa) {
        return new CompositeCell(2, [v, oa]);
    }
    static data(d) {
        return new CompositeCell(3, [d]);
    }
}

export function CompositeCell_$reflection() {
    return union_type("ARCtrl.CompositeCell", [], CompositeCell, () => [[["Item", OntologyAnnotation_$reflection()]], [["Item", string_type]], [["Item1", string_type], ["Item2", OntologyAnnotation_$reflection()]], [["Item", Data_$reflection()]]]);
}

